### ISSUE TITLE
[Errno 24] Too many open files

### ISSUE THREAD

**Issue by innjoshka** [MechanicalSoup/MechanicalSoup#204]
I encountered kinda bug with Mechanicalsoup relieving  socket connections. So, Begin:

This is my code:

```
import os
import mechanicalsoup
import time



url = "XXX"
file = os.path.join(os.getcwd(), 'data.csv')
counter = row = 30000

def get_data(file):
    with open(file, 'r') as myfile:
        data = myfile.read().splitlines()
        return data

data = get_data(file)


def fetchfile(query):
    browser = mechanicalsoup.StatefulBrowser(session=True, soup_config={'features': 'html.parser'})
    page = browser.get(url, timeout=10.0)
    search_form = page.soup.find("form", {"method": "POST"})
    search_form.find("input", {"name": "address"})["value"] = query
    response = browser.submit(search_form, page.url)
    return response



for i in range(10000000):
    counter = row = counter + 1
    print(counter)
    try:
        result = fetchfile(data[row])
        print(data[row], result)
    except Exception as e:
        print(e)
        time.sleep(15)
        pass

```

This is a simple code where I have row data and I make GET requests continuously (synchronously). So, during some time (min 15) I get error [Errno 24] Too many open files. I googled and found that this is closely linked with CLOSE_WAIT connections. on linux default number is 1024 `ulimit -n`. The idea is that Mechanicalsoup doesn't free connection even if it's closed by server.  So we'are going right into error of 1024 available connections working at once. and Also somehow Each iteration os loop Memory consumption always increasing by 1-2mb. Would be great that someone explain me why.

Screen showing what's happening with help of **netstat**
![screen shot 2018-04-06 at 11 50 43 am](https://user-images.githubusercontent.com/23149363/38412166-68d71168-3991-11e8-86ef-e98ffc28ef88.png)

![screen shot 2018-04-06 at 11 50 21 am](https://user-images.githubusercontent.com/23149363/38412250-987b31ba-3991-11e8-9396-cd8260dbd333.png)

By the way, all works fine on MacOS and Windows. The testing Linux server where bug happens is Ubuntu 16.04.4 LTS (GNU/Linux 4.4.0-1054-aws x86_64). As well as all works fine using just simple **requests** library with similar code.

 I need Mechanicalsoup works on Linux server, any suggestions what I have to try to make it work correctly?

ADDED: Library Versions 
MechanicalSoup version (0.10.0)
beautifulsoup4 (4.6.0)
requests (2.18.4)
urllib3 (1.22)
lxml (4.2.1)
python (3.5.2)

**Comment by innjoshka** [comment]
Thanks a lot for the help. But it still mention that I was able to get rid of the always creating and not closing new connections after server was closing it only when I moved 2 rows out of the loop and this is strange because it has to do it automatically as it happens using **requests**
```
browser = mechanicalsoup.StatefulBrowser(session=True, soup_config={'features': 'html.parser'})
page = browser.get(url, timeout=10.0)
```
So, I had to do a) move it b) close broswer manually, `browser.close()`. But anyway all works as I need. Thanks a lot. 

The only one question leaved is: how is better to use **asynchronous** requests with Mechanicalsoup? 
Better to use Asyncio , multiprocessing or grequests library?

**Comment by moy** [comment]
I can't help if you don't give the version of MechanicalSoup you're using ...

We made no effort towards asynchronous requests: the common use-case of MechanicalSoup is when the pages just fetched are used in the following actions (find a link in the page, ...), and it this case asynchronous requests are useless.

**Comment by innjoshka** [comment]
I've Edited issue , the end of the post: MechanicalSoup version (0.10.0)

**Comment by innjoshka** [comment]
Is it normal working behaviour?

When I assign browser outside the loop and try **open method** I get **error** `mechanicalsoup.utils.LinkNotFoundError`
```
browser = mechanicalsoup.StatefulBrowser(session=False, raise_on_404=True,
                                                 user_agent=headers['User-Agent'],
                                                 soup_config={'features': 'html.parser'})
def fetchfile(query):
    browser.open(self.url, verify = False, timeout=6.0)     
    browser.select_form('form[action="/submit"]')
    browser["address"] = query #by name
    browser.get_current_form().print_summary()
    response = browser.submit_selected()
    return response
```

But at the same time using **get method** all works fine 

```
browser = mechanicalsoup.StatefulBrowser(session=False, raise_on_404=True,
                                                 user_agent=headers['User-Agent'],
                                                 soup_config={'features': 'html.parser'})
def fetchfile(query):
    page = browser.get(self.url, timeout=6.0)
    search_form = page.soup.select('div.main-input')[0]
    search_form.select("input")[0]["value"] = query
    response = browser.submit(search_form, self.url)
    return response
```

**Comment by moy** [comment]
`open` is a very straightforward wrapper around `get`, so they should really have the same behavior.

You should try reproducing your issue on a public website like http://httpbin.org/ and post the exact code here so that we can reproduce it.

**Comment by hemberger** [comment]
There are a lot of ways that you can raise a `LinkNotFoundError` in MechanicalSoup, so it's possible that another line is causing the problem here. If you can provide the error message you're getting (or better yet, some reproducible code that we can run, as per @moy's suggestion), then we will be better equipped to help diagnose the issue.

**Comment by innjoshka** [comment]
I found what caused error **LinkNotFoundError** when tested it on [http://httpbin.org ](url) . It was because of site on what I tested the code do redirections, so it was immediately change the url to redirected one and because I tested the code in the loop - I got LinkNotFoundError. But didn't find the way how to turn redirection off when using browser.open(). browser.open(allow_redirects=False) didn't help to turn redirection off as well as not give me any errors. But I assume the issue can be closed. Thanks a lot for the help
